### PR TITLE
agent: use bind over advertise for src

### PR DESCRIPTION
**Pull request by magiconair**
This is a follow-up of #2822 which uses the bind address as source and not the advertise address since that won't work in NAT environments.

### CODE REVIEW

File: command/agent/config.go [hunk 1]
```diff
@@ -1976,3 +1976,21 @@ func (d dirEnts) Less(i, j int) bool {
 func (d dirEnts) Swap(i, j int) {
 	d[i], d[j] = d[j], d[i]
 }
+
+// inaddr_any checks if the given ip address is an IPv4 or IPv6 ANY address. ip
+// can be either a *net.IP or a string. It panics on another type.
+func inaddr_any(ip interface{}) bool {
```
Comment (slackpad) [inline command/agent/config.go]: This seems like an odd name with an underscore. Maybe `isAddrANY`?

File: command/agent/config.go [hunk 2]
```diff
@@ -1976,3 +1976,21 @@ func (d dirEnts) Less(i, j int) bool {
...
+	if ip == nil {
+		return false
+	}
+	var ips string
```
Comment (slackpad) [inline command/agent/config.go]: What's the path where a string gets passed in here?

File: command/agent/config.go [hunk 3]
```diff
@@ -1976,3 +1976,21 @@ func (d dirEnts) Less(i, j int) bool {
...
+	switch x := ip.(type) {
+	case *net.IP:
+		ips = x.String()
+	case string:
+		ips = x
+	default:
+		panic(fmt.Sprintf("invalid type: %T", ip))
+	}
+	return ips == "0.0.0.0" || ips == "::" || ips == "[::]"
```
Comment (slackpad) [inline command/agent/config.go]: Realized that there's https://golang.org/pkg/net/#IP.IsUnspecified which looks to be doing the same thing.

Comment (magiconair) [inline command/agent/config.go]: I suspected as much but couldn't find it. Thx

Comment (magiconair) [inline command/agent/config.go]: That doesn't work with `[::]`. I'll use my helper for now.

**Comment by allandrick** [comment]
@slackpad would you mind bumping the official-images docker library for this please?

**Comment by slackpad** [comment]
Hi @allandrick we are mid-cycle for 0.8.4 so this should be out in a couple weeks and we will bump it at that time. We currently don't update any existing images, we just release new ones as we release new Consul versions.

**Pull request merged**